### PR TITLE
hobbes: init at latest

### PR DESCRIPTION
**Pull request by thmzlt**
###### Motivation for this change

Introducing the [hobbes](morgan-stanley/hobbes) programming and execution environment build and maintained at Morgan Stanley.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thmzlt** [comment]
Looking into the build timeout. [It passes](https://travis-ci.org/Morgan-Stanley/hobbes/jobs/642605695) on Travis CI.

File: pkgs/development/tools/hobbes/default.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  doCheck = true;
+  checkTarget = "test";
+
+  meta = with stdenv.lib; {
+    description = "A language and an embedded JIT compiler";
+    longDescription = ''
+      Hobbes is a a language, embedded compiler, and runtime for efficient
+      dynamic expression evaluation, data storage and analysis.
+    '';
+    homepage = https://github.com/Morgan-Stanley/hobbes;
```
Comment (doronbehar) [inline pkgs/development/tools/hobbes/default.nix]:
```suggestion
    homepage = "https://github.com/Morgan-Stanley/hobbes";
```

File: pkgs/development/tools/hobbes/default.nix [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{ stdenv, pkgs }:
```
Comment (doronbehar) [inline pkgs/development/tools/hobbes/default.nix]: We usually import all the necessary arguments not with `pkgs` but with each package showing up in the list here.

File: pkgs/development/tools/hobbes/default.nix [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    cmake
+    gcc6
+    llvm_6
+    ncurses
+    python27
+    readline
+    zlib
+  ];
+
+  doCheck = true;
```
Comment (doronbehar) [inline pkgs/development/tools/hobbes/default.nix]: Didn't the builder ran `make check` or whatever without this?

File: pkgs/development/tools/hobbes/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  checkTarget = "test";
```
Comment (doronbehar) [inline pkgs/development/tools/hobbes/default.nix]: Usually this is used by default.

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/hobbes/default.nix (hunk 3 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: Good point. It probably didn't run because the checkTarget was incorrect.

File: pkgs/development/tools/hobbes/default.nix (hunk 4 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: If I removed it, the tests don't run. Isn't the default target "check"?

File: pkgs/development/tools/hobbes/default.nix (hunk 3 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: Builder didn't run the tests without this flag (even with the right target).

File: pkgs/development/tools/hobbes/default.nix (hunk 4 above)
Comment (doronbehar) [reply pkgs/development/tools/hobbes/default.nix]: Correct, my bad.

**Review by doronbehar — APPROVED** [review]

File: pkgs/development/tools/hobbes/default.nix [hunk 5]
```diff
@@ -0,0 +1,38 @@
+{ stdenv, fetchFromGitHub, cmake, gcc6, llvm_6, ncurses, python27, readline, zlib }:
+
+stdenv.mkDerivation {
+  name = "hobbes";
+  version = "latest-432d6a2";
```
Comment (doronbehar) [inline pkgs/development/tools/hobbes/default.nix]:
```suggestion
  version = "unstable-2020-02-18";
```

**Review by doronbehar — COMMENTED** [review]
Besides that all LGTM.

File: pkgs/development/tools/hobbes/default.nix (hunk 5 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: Done.

**Review by doronbehar — APPROVED** [review]

**Comment by thmzlt** [comment]
What else can I do to help with this PR? Do I need to tick all the boxes on the template list?

**Comment by doronbehar** [comment]
You can try and write at https://discourse.nixos.org/t/prs-already-reviewed/2617 . I don't have merge permissions so I've done as best as I could. Since I can see this is your first contribution, 1st of all welcome aboard :) And 2ndly, get used to it - PRs can take a longgg of time.

**Comment by thmzlt** [comment]
Thanks @doronbehar!

File: pkgs/development/tools/hobbes/default.nix [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  version = "unstable-2020-02-18";
+
+  src = fetchFromGitHub {
+    owner = "morgan-stanley";
+    repo = "hobbes";
+    rev = "8f920f320223f301a3e732e478a2facafd94687a";
+    sha256 = "0av6wp3gsjnil2j8xfn3rkx2r50sfcaq19d7a8bpnfhq2z8q67kg";
+  };
+
+  nativeBuildInputs = [
```
Comment (danieldk) [inline pkgs/development/tools/hobbes/default.nix]: I don't know this project, but `ncurses`, `readline`, and `zlib` should most likely be in `buildInputs`. From a quick glance at their `CMakeList.txt`, `llvm_6` should probably also be in`buildInputs`. If Python is only used during testing (as it appears), it should be in `checkInputs`. Also, did you try if this works with Python 3? Python 2 is EOL.

Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: Pushed changes reflecting your suggestions (and squashed the other commits). It does work with Python 3. Does it look ok?

Comment (danieldk) [reply pkgs/development/tools/hobbes/default.nix]: Did you try to put Python in `checkInputs`? Looks ok otherwise. Can't do a proper review though, since I can't test this package.

**Comment by aanderse** [comment]
@thmzlt after reading the upstream guide to building it seems like more recent `gcc` and `llvm` should work... did you try?
I believe @danieldk has a comment about `python` and `checkInputs` which hasn't been addressed yet.

**Comment by thmzlt** [comment]
> @thmzlt after reading the upstream guide to building it seems like more recent gcc and llvm should work... did you try?

Building with latest GCC and LLVM fails. There are open issues on upstream for that.

> I believe @danieldk has a comment about python and checkInputs which hasn't been addressed yet.

I'll fix it later today when I have a chance and let you know. Thank you all for guiding me on this PR.

**Comment by aanderse** [comment]
@thmzlt awesome! Links to upstream issues always appreciated so future updates can easily check if these issues are resolved or not.

File: pkgs/development/tools/hobbes/default.nix (hunk 6 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: Just did that. Indeed Python is only used to test the implementation of the fregion script.

**Comment by thmzlt** [comment]
Pushed changes to set Python as a check phase input. Should be good to go.

Regarding the issues related to LLVM/GCC versions, they are:

- https://github.com/Morgan-Stanley/hobbes/issues/315: LLVM is used as a JIT compiler and there are breaking changes in its API between releases
- https://github.com/Morgan-Stanley/hobbes/issues/316: Warnings (that fail the build) are triggered by GCC 8

Both issues are expected to be fixed.

**Comment by danieldk** [comment]
> Pushed changes to set Python as a check phase input. Should be good to go.

Cool, thanks!
 
> Regarding the issues related to LLVM/GCC versions, they are:

Maybe it's worthwhile adding a short comment to the derivation above the respective dependencies with the links? Avoid that somebody has to hunt for PRs in the future to find these links ;).

File: pkgs/development/tools/hobbes/default.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    cmake
+    gcc6
```
Comment (aanderse) [inline pkgs/development/tools/hobbes/default.nix]: Should be removed, see above.

**Review by aanderse — CHANGES_REQUESTED** [review]
I notice there are a few unresolved comments from other reviewers.

**Comment by thmzlt** [comment]
@aanderse, just pushed a new version of the derivation with your feedback. Let me know if there's anything else.

File: pkgs/development/tools/hobbes/default.nix (hunk 5 above)
Comment (aanderse) [reply pkgs/development/tools/hobbes/default.nix]: I don't see this as being done yet.

File: pkgs/development/tools/hobbes/default.nix [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    ncurses
+    readline
+    zlib
+  ];
+
+  checkInputs = [
+    python37
+  ];
+
+  doCheck = false;
```
Comment (aanderse) [inline pkgs/development/tools/hobbes/default.nix]: I thought `python` was used for tests but now this is disabled.

File: maintainers/maintainer-list.nix [hunk 9]
```diff
@@ -3976,6 +3976,12 @@
     githubId = 10544;
     name = "Giuluo Eulisse";
   };
+  kthielen = {
```
Comment (aanderse) [inline maintainers/maintainer-list.nix]: Please separate this into a separate commit. So I'm asking for you to add yourself to the maintainer list in the forest commit, followed by the rest of this PR in your second commit.

**Review by aanderse — CHANGES_REQUESTED** [review]
This is looking really good! Very close, a few more changes, a quick test, and I'm more than happy to merge. Thanks for your patience up to this point.

File: pkgs/development/tools/hobbes/default.nix (hunk 5 above)
Comment (thmzlt) [reply pkgs/development/tools/hobbes/default.nix]: My bad, fixing it right away.

**Comment by thmzlt** [comment]
@aanderse, I have fixed the version name, and separated the changes in two commits. I have also changed the maintiner from the upstream's maintainer to myself since I'm the one working on the derivation. Finally, I have added a comment pointing to the upstream issue on why doCheck is currently set to false, and removed the python37 input.

**Review by doronbehar — APPROVED** [review]

**Review by aanderse — APPROVED** [review]
@thmzlt what a fantastic job you've done here. Welcome to the maintainers team, we're so glad to have you! :tada:

**Comment by aanderse** [comment]
@GrahamcOfBorg build hobbes

**Comment by aanderse** [comment]
Oh no! :scream: `x86_64-darwin` failure!

**Comment by thmzlt** [comment]
I can't test it on Darwin at the moment (I don't have a supported computer). Is it possible to mark this as Linux-only while I figure out how to build on a Mac?

**Comment by doronbehar** [comment]
Yes. If upstream intended this to build on darwin, then we mark this as broken there. I.e use `meta.broken = stdenv.isDarwin;`.

If upstream intended this to work only on Linux, you can use the `platforms` meta attribute.

**Comment by thmzlt** [comment]
@GrahamcOfBorg build hobbes

**Review by doronbehar — APPROVED** [review]

**Comment by thmzlt** [comment]
Looks like we are good. Thank you all for the help.
